Interestingly, gmoccapy, qtdragon and touchy do not exit on SIGTERM. The output contains:
```
WARN: linuxcnc did not exit on SIGTERM, escalating to KILL
```
Wonder why that is.

Interesting warning:
```
[Gmoccapy][WARNING]  No virtual keyboard installed, we checked for <onboard>. Try 'sudo apt-get install onboard'. (gmoccapy:2311)
```
Should we try to have a virtual keyboard? Or is this just, well, don't care? Probably not usable anyway.

This one may be interesting:
```
[Gmoccapy][ERROR]  Found an error!
The following information may be useful in troubleshooting:
Traceback (most recent call last):
  File "/.../grandixximo/lib/python/gladevcp/hal_bar.py", line 232, in expose
    def expose(self, widget, event):
    
KeyboardInterrupt
 (gmoccapy:62)
```

This one is present in multiple:
```
/.../grandixximo/lib/python/gladevcp/gtk_action.py:312: SyntaxWarning: 'return' in a 'finally' block
  return npath
```

The touchy test shows this:
```
* * * * * * * * * * - - - - - WARNING - - - - - * * * * * * * * * *
*    CMS_DISPLAY_ASCII_UPDATER may not function properly due      *
*    to range limitations of some of the update() functions.      *
* * * * * * * * * * - - - - - WARNING - - - - - * * * * * * * * * *
```
It may indicate that the NML file is not found or a wrong one is used.

Qtdragon and touchy output is also riddled with warnings. And a deprecation warning in touchy:
```
(touchy:2765547): Gtk-WARNING **: 13:31:04.022: Theme parsing error: <data>:1:26: Junk at end of value for font-family
/.../grandixximo/bin/touchy:651: DeprecationWarning: Gtk.Widget.override_font is deprecated
  w.override_font(self.control_font)
```

--- a/tests/ui-smoke/_lib/drive.py
+++ b/tests/ui-smoke/_lib/drive.py
@@ -13,6 +13,7 @@
 
 import argparse
 import linuxcnc
+import os
 import sys
 import time
 
@@ -32,6 +33,55 @@ ENSURE_ATTEMPT_TIMEOUT_S = 3.0
 STATE_STABILITY_S = 0.5
 STATE_RETRY_BUDGET = 6
 
+# linuxcnc launcher PID, written to linuxcnc.pid by the launcher and read
+# once at startup. The driver watches it so a GUI crash, which tears
+# linuxcnc down, fails the test in ~1s with a clear message instead of
+# waiting out a long NML poll. A dead task keeps serving its last stat
+# buffer, so process liveness is the only reliable crash signal.
+_WATCH_PID = None
+
+
+class LauncherGone(Exception):
+    """linuxcnc process group exited (GUI crashed or task died)."""
+
+
+def _read_pid(path):
+    try:
+        with open(path) as f:
+            return int(f.read().strip())
+    except (OSError, ValueError):
+        return None
+
+
+# Crash markers faulthandler and scripts/linuxcnc write to linuxcnc.err
+# the instant the GUI dies. The launcher PID can linger in Cleanup, so
+# scanning these catches the crash sooner and regardless of which GUI.
+_CRASH_MARKERS = ("Fatal Python error", "Segmentation fault", "Aborted")
+
+
+def _crash_marker_seen():
+    try:
+        with open("linuxcnc.err") as f:
+            return any(m in f.read() for m in _CRASH_MARKERS)
+    except OSError:
+        return False
+
+
+def _watchdog():
+    """Raise LauncherGone if the GUI has crashed: either the launcher PID
+    is gone, or a crash marker appeared in linuxcnc.err. Unknown PID and
+    a missing log count as alive, so a not-yet-written file never
+    false-fails the test."""
+    if _WATCH_PID is not None:
+        try:
+            os.kill(_WATCH_PID, 0)
+        except ProcessLookupError:
+            raise LauncherGone()
+        except PermissionError:
+            pass
+    if _crash_marker_seen():
+        raise LauncherGone()
+
 
 def connect_and_wait_ready(timeout):
     """Wait until linuxcnc.stat().poll() returns without error and
@@ -47,6 +97,7 @@ def connect_and_wait_ready(timeout):
     deadline = time.monotonic() + timeout
     last_err = None
     while time.monotonic() < deadline:
+        _watchdog()
         try:
             stat = linuxcnc.stat()
             stat.poll()
@@ -70,6 +121,7 @@ def wait_until_quiet(stat, predicate, timeout):
     must not happen."""
     deadline = time.monotonic() + timeout
     while time.monotonic() < deadline:
+        _watchdog()
         stat.poll()
         if predicate(stat):
             return True
@@ -195,6 +247,7 @@ def wait_program_started(stat, timeout):
     IDLE; we then read stat.position at (0,0,0)."""
     deadline = time.monotonic() + timeout
     while time.monotonic() < deadline:
+        _watchdog()
         stat.poll()
         if stat.interp_state != linuxcnc.INTERP_IDLE:
             return True
@@ -214,6 +267,7 @@ def wait_program_idle(stat, timeout):
     deadline = time.monotonic() + timeout
     consecutive = 0
     while time.monotonic() < deadline:
+        _watchdog()
         stat.poll()
         idle = (
             stat.interp_state == linuxcnc.INTERP_IDLE
@@ -311,29 +365,40 @@ def main():
     if args.run_program and args.expect_delta_mm is None:
         ap.error("--run-program requires --expect-delta-mm DX,DY,DZ")
 
-    cmd, stat = connect_and_wait_ready(CONNECT_TIMEOUT_S)
-    if cmd is None:
-        return 1
+    global _WATCH_PID
+    _WATCH_PID = _read_pid("linuxcnc.pid")
 
-    # Give the GUI process enough time to finish constructing itself
-    # (load .ui files, compile resources.py if needed, etc.) and
-    # settle. If the GUI was going to crash on startup it has crashed
-    # by now.
-    time.sleep(SETTLE_S)
-
-    # Re-check task is still alive; a GUI crash may have torn linuxcnc
-    # down via Cleanup.
     try:
-        stat.poll()
-    except linuxcnc.error as e:
-        sys.stderr.write(f"UI_SMOKE_FAIL: task disappeared after GUI startup: {e}\n")
-        return 1
-
-    if args.run_program:
-        if not run_program(cmd, stat,
-                           args.run_program, args.expect_delta_mm,
-                           args.tol, args.run_timeout):
+        cmd, stat = connect_and_wait_ready(CONNECT_TIMEOUT_S)
+        if cmd is None:
             return 1
+
+        # Give the GUI process enough time to finish constructing itself
+        # (load .ui files, compile resources.py if needed, etc.) and
+        # settle. If the GUI was going to crash on startup it has crashed
+        # by now.
+        time.sleep(SETTLE_S)
+        _watchdog()
+
+        # Re-check task is still alive; a GUI crash may have torn linuxcnc
+        # down via Cleanup.
+        try:
+            stat.poll()
+        except linuxcnc.error as e:
+            sys.stderr.write(f"UI_SMOKE_FAIL: task disappeared after GUI startup: {e}\n")
+            return 1
+
+        if args.run_program:
+            if not run_program(cmd, stat,
+                               args.run_program, args.expect_delta_mm,
+                               args.tol, args.run_timeout):
+                return 1
+    except LauncherGone:
+        sys.stderr.write(
+            "UI_SMOKE_FAIL: linuxcnc exited before the driver finished; "
+            "the GUI crashed or task died. See linuxcnc.out / linuxcnc.err "
+            "above for the backtrace.\n")
+        return 1
 
     print("UI_SMOKE_OK")
     return 0

--- a/tests/ui-smoke/_lib/launch-env.sh
+++ b/tests/ui-smoke/_lib/launch-env.sh
@@ -24,3 +24,8 @@ export CANBERRA_DRIVER=null
 export GST_PLUGIN_FEATURE_RANK="pulsesink:NONE,alsasink:NONE,osssink:NONE,oss4sink:NONE,jackaudiosink:NONE,pipewiresink:NONE,openalsink:NONE"
 export PULSE_SERVER=/dev/null
 export SDL_AUDIODRIVER=dummy
+
+# Dump a Python traceback on a fatal signal. For a pure-Python crash this
+# names the line; for a C/C++ crash (Qt, dbus, GL) it shows the Python
+# frame that called in. The native side is captured by crashdump.sh.
+export PYTHONFAULTHANDLER=1

--- a/tests/ui-smoke/_lib/run-gui.sh
+++ b/tests/ui-smoke/_lib/run-gui.sh
@@ -12,8 +12,17 @@ LIB_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
 TEST_DIR="${TEST_DIR:-$(cd "$(dirname "$0")" && pwd)}"
 CONFIGS_DIR="$(cd "$LIB_DIR/../../../configs/sim" && pwd)"
 
-INI_REL="$1"
+INI_ARG="$1"
 shift
 
+# Accept either a relative path under configs/sim/ or an absolute path.
+# Absolute paths are used by tests that need to point at a writable
+# mirror of a shipped config (qtdragon writes a log file inside the
+# config dir, which is read-only on CI).
+case "$INI_ARG" in
+    /*) INI_PATH="$INI_ARG" ;;
+    *)  INI_PATH="$CONFIGS_DIR/$INI_ARG" ;;
+esac
+
 export TEST_DIR
-exec "$LIB_DIR/launch.sh" "$CONFIGS_DIR/$INI_REL" "$@"
+exec "$LIB_DIR/launch.sh" "$INI_PATH" "$@"

--- a/tests/ui-smoke/qtdragon/test.sh
+++ b/tests/ui-smoke/qtdragon/test.sh
@@ -1,4 +1,21 @@
 #!/bin/bash
+# qtdragon's qtvcp logger writes its log file (path from INI [DISPLAY]
+# LOG_FILE) into the config directory. CI mounts the workspace read-
+# only for the runtime user, so a relative LOG_FILE like 'qtdragon.log'
+# resolves to a path qtvcp cannot create, hal_bridge then exits, and
+# linuxcnc tears down before our driver can do anything. Mirror the
+# config dir to a writable tmp location and patch LOG_FILE to be
+# rooted at $HOME so the log lands in a directory we can write to.
+set -u
+
 LIB_DIR="$(cd "$(dirname "$0")/../_lib" && pwd)"
-exec "$LIB_DIR/run-gui.sh" qtdragon/qtdragon_xyz/qtdragon_metric.ini \
+SRC_DIR="$(cd "$LIB_DIR/../../../configs/sim/qtdragon/qtdragon_xyz" && pwd)"
+
+WORK_DIR="$(mktemp -d -t ui-smoke-qtdragon.XXXXXX)"
+trap 'rm -rf "$WORK_DIR"' EXIT
+cp -r "$SRC_DIR/." "$WORK_DIR/"
+sed -i 's|^LOG_FILE = qtdragon\.log$|LOG_FILE = ~/qtdragon.log|' \
+    "$WORK_DIR/qtdragon_metric.ini"
+
+exec "$LIB_DIR/run-gui.sh" "$WORK_DIR/qtdragon_metric.ini" \
     --run-program "$LIB_DIR/smoke.ngc" --expect-delta-mm 1,1,0

--- a/tests/ui-smoke/qtdragon/test.sh
+++ b/tests/ui-smoke/qtdragon/test.sh
@@ -15,6 +15,12 @@
 # scripts/linuxcnc itself forces QT_QPA_PLATFORM=xcb unless
 # LINUXCNC_OPENGL_PLATFORM is set to something other than glx, so we
 # pin both env vars.
+#
+# qtdragon embeds a QWebEngineView (Chromium). Under offscreen + xvfb
+# with no GPU and no user namespaces in the CI runner sandbox,
+# QtWebEngine racy-crashes during browser-process spawn. Disable the
+# Chromium sandbox and force single-process + software rendering so
+# the renderer runs in the same process as Qt with no GPU thread.
 set -u
 
 LIB_DIR="$(cd "$(dirname "$0")/../_lib" && pwd)"
@@ -28,6 +34,8 @@ sed -i 's|^LOG_FILE = qtdragon\.log$|LOG_FILE = ~/qtdragon.log|' \
 
 export LINUXCNC_OPENGL_PLATFORM=offscreen
 export QT_QPA_PLATFORM=offscreen
+export QTWEBENGINE_DISABLE_SANDBOX=1
+export QTWEBENGINE_CHROMIUM_FLAGS="--no-sandbox --disable-gpu --disable-software-rasterizer --single-process --no-zygote"
 
 exec "$LIB_DIR/run-gui.sh" "$WORK_DIR/qtdragon_metric.ini" \
     --run-program "$LIB_DIR/smoke.ngc" --expect-delta-mm 1,1,0

